File: sdks/scala/core/js/src/main/scala/golem/host/QuotaApi.scala
/*
 * Copyright 2024-2026 John A. De Goes and the ZIO Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golem.host

import golem.host.js._
import golem.schema.{
  FromSchema,
  FromSchemaError,
  GuestQuotaTokenHandle,
  IntoSchema,
  QuotaTokenSpec,
  SchemaGraph,
  SchemaType,
  SchemaTypeBody,
  SchemaValue
}

import scala.collection.immutable.ListMap
import scala.concurrent.{ExecutionContext, Future}
import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport

/**
 * Scala.js facade for `golem:quota/types@1.5.0`.
 *
 * The `quota-token` capability is an opaque, unforgeable, owned resource
 * defined in `golem:core/types@2.0.0`. Guest code can only hold and move the
 * handle: it cannot read the token's internals, serialize it, or forge one. The
 * quota interface exposes free functions over such a handle plus the
 * `reservation` resource:
 * {{{
 *   record failed-reservation { estimated-wait-nanos: option<u64> }
 *   resource reservation { commit: static func(this: reservation, used: u64) }
 *   new-token: func(resource-name: string, expected-use: u64) -> quota-token;
 *   reserve:   func(token: borrow<quota-token>, amount: u64)
 *                -> result<reservation, failed-reservation>;
 *   split:     func(token: borrow<quota-token>, child-expected-use: u64) -> quota-token;
 *   merge:     func(token: borrow<quota-token>, other: quota-token);
 * }}}
 */
object QuotaApi {

  // --- WIT: failed-reservation record ---

  /**
   * Returned when a reservation cannot be granted (enforcement policy =
   * `reject`).
   */
  final case class FailedReservation(
    /**
     * Estimated wait time in nanoseconds; only set for rate-limited resources.
     */
    estimatedWaitNanos: Option[BigInt]
  )

  // --- WIT: reservation resource ---

  /**
   * A short-lived capability representing a pending resource consumption.
   *
   * Dropping without calling [[commit]] is equivalent to committing zero usage.
   */
  final class Reservation private[golem] (private[golem] val underlying: JsReservation) {

    /**
     * Commit actual usage.
     *   - `used` < reserved → unused capacity is returned to the pool.
     *   - `used` > reserved → excess is deducted as "debt" from the token's
     *     allocation.
     */
    def commit(used: BigInt): Unit =
      JsReservationStatic.commit(underlying, js.BigInt(used.toString))
  }

  // --- WIT: quota-token resource ---

  /**
   * An unforgeable capability granting the right to consume a named resource.
   *
   * A `QuotaToken` holds only an opaque, affine handle to the owned host
   * resource: it carries no readable content and can be transferred to exactly
   * one destination (an RPC argument, a method return value, ...). Once
   * transferred the token can no longer be used; [[split]] first if you need to
   * both keep and send a capability.
   *
   * Typical usage:
   * {{{
   *   val token = QuotaToken("openai-tokens", BigInt(1000))
   *   token.reserve(BigInt(500)) match {
   *     case Right(reservation) =>
   *       val used = callExternalApi()
   *       reservation.commit(used)
   *     case Left(err) =>
   *       println(s"quota unavailable: $err")
   *   }
   * }}}
   */
  final class QuotaToken private[golem] (private[golem] val handle: GuestQuotaTokenHandle) {

    /**
     * Reserve `amount` units from the local allocation.
     *
     * Blocks internally until capacity is available or the resource's
     * enforcement action fires. Returns `Right(reservation)` on success, or
     * `Left(FailedReservation)` when the enforcement policy is `reject`. For
     * `throttle` / `terminate` policies the call suspends or terminates the
     * agent before returning.
     *
     * Traps if this token has already been transferred.
     */
    def reserve(amount: BigInt): Either[FailedReservation, Reservation] =
      handle.withHandle { raw =>
        try
          Right(
            new Reservation(QuotaModule.reserve(raw.asInstanceOf[JsQuotaTokenResource], js.BigInt(amount.toString)))
          )
        catch {
          case e: js.JavaScriptException =>
            val rawErr        = e.exception.asInstanceOf[JsFailedReservation]
            val estimatedWait = rawErr.estimatedWaitNanos.toOption.map(bi => BigInt(bi.toString))
            Left(FailedReservation(estimatedWait))
        }
      }
        .getOrElse(throw new IllegalStateException(TOKEN_CONSUMED))

    /**
     * Split off a child token with `childExpectedUse` units of expected-use.
     *
     *   - The parent's expected-use is reduced by `childExpectedUse`.
     *   - Credits are divided proportionally between parent and child.
     *
     * Traps if `childExpectedUse` exceeds the parent's current expected-use, or
     * if this token has already been transferred.
     */
    def split(childExpectedUse: BigInt): QuotaToken =
      handle.withHandle { raw =>
        new QuotaToken(
          GuestQuotaTokenHandle.fromRaw(
            QuotaModule.split(raw.asInstanceOf[JsQuotaTokenResource], js.BigInt(childExpectedUse.toString))
          )
        )
      }
        .getOrElse(throw new IllegalStateException(TOKEN_CONSUMED))

    /**
     * Merge `other` back into this token.
     *
     * Combines expected-use and credits. `other` is consumed by this call and
     * must not be used afterwards; this token remains usable.
     *
     * Traps if the tokens refer to different resources, or if either token has
     * already been transferred.
     */
    def merge(other: QuotaToken): Unit = {
      // Reject merging a token into itself before taking any handle, so a shared
      // handle is not consumed by the receiver and then read again as `other`.
      if (other.handle eq this.handle)
        throw new IllegalArgumentException("cannot merge a quota token with itself")
      // Borrow this token first (it is a `borrow<quota-token>` and stays usable);
      // a consumed receiver must not consume `other`.
      handle.withHandle { thisRaw =>
        other.handle.take() match {
          case Some(otherRaw) =>
            QuotaModule.merge(
              thisRaw.asInstanceOf[JsQuotaTokenResource],
              otherRaw.asInstanceOf[JsQuotaTokenResource]
            )
          case None => throw new IllegalStateException(TOKEN_CONSUMED)
        }
      }
        .getOrElse(throw new IllegalStateException(TOKEN_CONSUMED))
    }

    /**
     * Reserve `amount` units, run `f`, then commit the actual usage returned by
     * `f`. Commits zero usage on failure and re-throws.
     */
    def withReservation[T](amount: BigInt)(
      f: Reservation => Future[(BigInt, T)]
    ): Future[Either[FailedReservation, T]] =
      QuotaApi.withReservation(this, amount)(f)
  }

  object QuotaToken {

    /**
     * Request a quota capability for the named resource.
     *
     * @param resourceName
     *   The resource name as declared in the manifest.
     * @param expectedUse
     *   Expected units per reservation; used to derive the credit rate and
     *   max-credit for fair scheduling.
     */
    def apply(resourceName: String, expectedUse: BigInt): QuotaToken =
      new QuotaToken(
        GuestQuotaTokenHandle.fromRaw(QuotaModule.newToken(resourceName, js.BigInt(expectedUse.toString)))
      )

    /**
     * Automatic serialization for RPC: a `QuotaToken` is a schema-native
     * capability node (`golem:core/types@2.0.0` `quota-token` /
     * `quota-token-handle`), not a plain record. The owned handle is moved into
     * the value tree when it is encoded, so a token can be sent exactly once
     * and cannot be forged from data.
     */
    implicit val intoSchema: IntoSchema[QuotaToken] =
      new IntoSchema[QuotaToken] {
        override lazy val graph: SchemaGraph =
          SchemaGraph(ListMap.empty, SchemaType(SchemaTypeBody.QuotaTokenType(QuotaTokenSpec())))

        override def toValue(token: QuotaToken): SchemaValue =
          SchemaValue.QuotaTokenHandle(token.handle)
      }

    implicit val fromSchema: FromSchema[QuotaToken] =
      new FromSchema[QuotaToken] {
        override def fromValue(value: SchemaValue): Either[FromSchemaError, QuotaToken] =
          value match {
            case SchemaValue.QuotaTokenHandle(h) => Right(new QuotaToken(h))
            case other                           =>
              Left(FromSchemaError(s"expected quota-token handle for QuotaToken, got $other"))
          }
      }
  }

  private val TOKEN_CONSUMED =
    "quota token has already been transferred and can no longer be used; split the token first if " +
      "you need to both keep and send a capability"

  private implicit val ec: ExecutionContext = ExecutionContext.global

  /**
   * Reserve `amount` units from `token`, run `f`, then commit the actual usage
   * returned by `f`. Commits zero usage on failure and re-throws, ensuring
   * unused capacity is always returned to the pool.
   *
   * Returns `Future(Left(FailedReservation))` if the reservation could not be
   * granted, or `Future(Right(value))` on success.
   *
   * {{{
   *   val result = withReservation(token, BigInt(500)) { reservation =>
   *     Future {
   *       val data = callExternalApi()
   *       (data.tokensUsed, data)
   *     }
   *   }
   * }}}
   */
  def withReservation[T](token: QuotaToken, amount: BigInt)(
    f: Reservation => Future[(BigInt, T)]
  ): Future[Either[FailedReservation, T]] =
    token.reserve(amount) match {
      case Left(err)          => Future.successful(Left(err))
      case Right(reservation) =>
        f(reservation).map { case (used, value) =>
          reservation.commit(used)
          Right(value)
        }.recoverWith { case e: Throwable =>
          reservation.commit(BigInt(0))
          Future.failed(e)
        }
    }

  @js.native
  @JSImport("golem:quota/types@1.5.0", JSImport.Namespace)
  private object QuotaModule extends js.Object {
    def newToken(resourceName: String, expectedUse: js.BigInt): JsQuotaTokenResource          = js.native
    def reserve(token: JsQuotaTokenResource, amount: js.BigInt): JsReservation                = js.native
    def split(token: JsQuotaTokenResource, childExpectedUse: js.BigInt): JsQuotaTokenResource = js.native
    def merge(token: JsQuotaTokenResource, other: JsQuotaTokenResource): Unit                 = js.native
  }

  @js.native
  @JSImport("golem:quota/types@1.5.0", "Reservation")
  private object JsReservationStatic extends js.Object {
    def commit(self: JsReservation, used: js.BigInt): Unit = js.native
  }
}

File: sdks/kotlin/sdk/src/wasmWasiMain/kotlin/cloud/golem/runtime/host/QuotaApi.kt
@file:OptIn(kotlin.wasm.unsafe.UnsafeWasmMemoryApi::class, kotlin.wasm.ExperimentalWasmInterop::class)

package cloud.golem.runtime.host

import cloud.golem.runtime.Either
import cloud.golem.wasm.alloc
import cloud.golem.wasm.loadByte
import cloud.golem.wasm.loadInt
import cloud.golem.wasm.loadLong
import cloud.golem.wasm.storeByte

// Raw canonical-ABI bindings to golem:quota/types@1.5.0. The `quota-token` capability itself is
// golem:core/types@2.0.0's `quota-token` resource -- the SAME resource
// (SchemaValue.kt's SecretVal/QuotaTokenVal) already established has NO methods of its own; this
// interface only exposes free functions that act on a `quota-token` handle, plus the
// `reservation` resource. Signatures verified via abi-dump's `sig`/`resulttype` modes against
// wit-native/deps/golem-quota/types.wit.
//
// New wrinkle vs everything built so far: `reservation.commit` is declared
// `static func(this: reservation, used: u64)` -- a static function that takes the resource by
// VALUE (owned, not borrowed) as an explicit `this` parameter, not a `[method]`. Calling it
// consumes the reservation (the host releases the handle as part of the call); there is no
// separate [resource-drop]reservation call needed after a successful commit. Same
// `[static]<resource>.<name>` raw name shape as `[static]bucket.open-bucket`, just with
// the resource threaded as a normal parameter instead of being implicit.
@kotlin.wasm.WasmImport("golem:quota/types@1.5.0", "new-token")
private external fun hostNewToken(namePtr: Int, nameLen: Int, expectedUse: Long): Int

@kotlin.wasm.WasmImport("golem:quota/types@1.5.0", "reserve")
private external fun hostReserve(tokenHandle: Int, amount: Long, retPtr: Int)

@kotlin.wasm.WasmImport("golem:quota/types@1.5.0", "split")
private external fun hostSplit(tokenHandle: Int, childExpectedUse: Long): Int

@kotlin.wasm.WasmImport("golem:quota/types@1.5.0", "merge")
private external fun hostMerge(tokenHandle: Int, otherHandle: Int)

@kotlin.wasm.WasmImport("golem:quota/types@1.5.0", "[static]reservation.commit")
private external fun hostReservationCommit(reservationHandle: Int, used: Long)

private const val TOKEN_CONSUMED =
    "quota token has already been transferred and can no longer be used; split the token first " +
        "if you need to both keep and send a capability"

private fun lowerStringToPtrLen(s: String): Pair<Int, Int> {
    val bytes = s.encodeToByteArray()
    val ptr = alloc(bytes.size, 1)
    for (i in bytes.indices) storeByte(ptr + i, bytes[i])
    return ptr to bytes.size
}

/** Returned when a reservation cannot be granted (enforcement policy = `reject`). */
data class FailedReservation(val estimatedWaitNanos: Long?)

/**
 * A short-lived capability representing a pending resource consumption. Dropping without
 * calling [commit] is equivalent to committing zero usage -- this SDK doesn't expose a manual
 * drop for it (matching the Scala reference), only [commit], which itself consumes the handle.
 */
class Reservation internal constructor(handle: Int) {
    private var handle: Int? = handle

    /**
     * Commit actual usage. `used` < reserved returns unused capacity to the pool; `used` >
     * reserved deducts the excess as "debt" from the token's allocation. Consumes this
     * reservation -- calling [commit] twice fails.
     */
    fun commit(used: Long) {
        val h = handle ?: error("reservation already committed")
        handle = null
        hostReservationCommit(h, used)
    }
}

/**
 * An unforgeable capability granting the right to consume a named resource. Holds only an
 * opaque, affine handle to the owned host resource: it carries no readable content and can be
 * transferred to exactly one destination ([merge]'s `other` argument, or wherever this SDK's
 * schema-value-tree machinery sends a [cloud.golem.runtime.SchemaValue.QuotaTokenVal]). Once
 * transferred, the token can no longer be used -- [split] first if you need to both keep and
 * send a capability.
 */
class QuotaToken internal constructor(handle: Int) {
    private var handle: Int? = handle

    private fun <T> withHandle(block: (Int) -> T): T = block(handle ?: error(TOKEN_CONSUMED))

    /** Takes ownership of this token's handle for a one-time transfer (e.g. into [merge]'s `other`). Returns null if already consumed. */
    internal fun take(): Int? {
        val h = handle
        handle = null
        return h
    }

    /**
     * Reserve `amount` units from the local allocation. Blocks internally until capacity is
     * available or the resource's enforcement action fires. Returns `Right(reservation)` on
     * success, or `Left(FailedReservation)` when the enforcement policy is `reject`. Traps (via
     * [error]) if this token has already been transferred.
     */
    fun reserve(amount: Long): Either<FailedReservation, Reservation> = withHandle { h ->
        // result<reservation, failed-reservation>: tag@0(1,1), payload@8 (max(reservation i32/4,
        // failed-reservation{option<u64>} 16/8) = 16) -> 24 total, align8.
        val retPtr = alloc(24, 8)
        hostReserve(h, amount, retPtr)
        if (loadByte(retPtr).toInt() == 0) {
            Either.Right(Reservation(loadInt(retPtr + 8)))
        } else {
            val errBase = retPtr + 8 // failed-reservation: {estimated-wait-nanos: option<u64>} @0
            val hasWait = loadByte(errBase).toInt() != 0
            Either.Left(FailedReservation(if (hasWait) loadLong(errBase + 8) else null))
        }
    }

    /**
     * Split off a child token with `childExpectedUse` units of expected-use. The parent's
     * expected-use is reduced by `childExpectedUse`; credits are divided proportionally. Traps
     * if `childExpectedUse` exceeds the parent's current expected-use, or if this token has
     * already been transferred.
     */
    fun split(childExpectedUse: Long): QuotaToken = withHandle { h -> QuotaToken(hostSplit(h, childExpectedUse)) }

    /**
     * Merge `other` back into this token: combines expected-use and credits. `other` is
     * consumed by this call and must not be used afterwards; this token remains usable. Traps
     * if the tokens refer to different resources, or if either token has already been
     * transferred.
     */
    fun merge(other: QuotaToken) {
        require(other !== this) { "cannot merge a quota token with itself" }
        withHandle { thisHandle ->
            val otherHandle = other.take() ?: error(TOKEN_CONSUMED)
            hostMerge(thisHandle, otherHandle)
        }
    }

    /** Reserve `amount` units, run [block], then commit the actual usage [block] returns. Commits zero usage and rethrows on failure. */
    fun <T> withReservation(amount: Long, block: (Reservation) -> Pair<Long, T>): Either<FailedReservation, T> = QuotaApi.withReservation(this, amount, block)

    companion object {
        /**
         * Request a quota capability for the named resource.
         *
         * @param resourceName the resource name as declared in the manifest.
         * @param expectedUse expected units per reservation; used to derive the credit rate and max-credit for fair scheduling.
         */
        fun create(resourceName: String, expectedUse: Long): QuotaToken {
            val (namePtr, nameLen) = lowerStringToPtrLen(resourceName)
            return QuotaToken(hostNewToken(namePtr, nameLen, expectedUse))
        }
    }
}

object QuotaApi {
    /**
     * Reserve `amount` units from `token`, run [block], then commit the actual usage [block]
     * returns. Commits zero usage and rethrows on failure, ensuring unused capacity is always
     * returned to the pool.
     */
    fun <T> withReservation(token: QuotaToken, amount: Long, block: (Reservation) -> Pair<Long, T>): Either<FailedReservation, T> = when (val r = token.reserve(amount)) {
        is Either.Left -> Either.Left(r.value)
        is Either.Right -> {
            val reservation = r.value
            try {
                val (used, value) = block(reservation)
                reservation.commit(used)
                Either.Right(value)
            } catch (e: Throwable) {
                reservation.commit(0L)
                throw e
            }
        }
    }
}
